# USPS Data Prediction Using Daimensions

This dataset is from OpenML who describes the data as, "Normalized handwritten digits, automatically scanned from envelopes by the U.S. Postal Service."

## 0. Setup

We'll get the csv from the OpenML link and use a pandas dataframe to split it into training and validation data in csv's.

In [1]:
# using pandas to get csv as a dataframe and see how it looks
import pandas as pd
from sklearn.model_selection import train_test_split

dataset_url = 'https://www.openml.org/data/get_csv/19329737/usps.csv'
data = pd.read_csv(dataset_url)
data.head()

int0   double1   double2   double3   double4   double5   double6   double7  \
0     7 -1.000000 -1.000000 -1.000000 -0.999997 -0.999729 -0.989537 -0.860165   
1     6 -0.999927 -0.993644 -0.900309 -0.632621 -0.443145 -0.454436 -0.474872   
2     5 -1.000000 -1.000000 -1.000000 -0.999995 -0.999986 -0.999987 -0.999990   
3     8 -0.999998 -0.999800 -0.994136 -0.932288 -0.673685 -0.192056  0.218820   
4     4 -1.000000 -0.999971 -0.997452 -0.957499 -0.801681 -0.510623 -0.107965   

    double8   double9  ...  double247  double248  double249  double250  \
0 -0.395924  0.048299  ...   0.131298   0.384728   0.433073   0.193397   
1 -0.431176 -0.494539  ...  -0.388084  -0.273639   0.001094   0.294825   
2 -0.999316 -0.976299  ...  -0.999778  -0.988816  -0.821857  -0.277379   
3  0.337224  0.193371  ...  -0.373861   0.196530   0.002823  -0.695900   
4  0.217262  0.216431  ...   0.214362   0.339240   0.458489   0.473456   

   double251  double252  double253  double254  double255  double256  
0  -0.237439  -0.649848  -0.906359  -0.988672  -0.999475  -0.999994  
1   0.316915   0.113494  -0.092331  -0.329390  -0.584681  -0.868793  
2   0.136501  -0.247816  -0.809498  -0.988116  -0.999864  -1.000000  
3  -0.976372  -0.999458  -0.999996  -1.000000  -1.000000  -1.000000  
4   0.372240   0.173322  -0.144182  -0.542254  -0.863031  -0.981978  

[5 rows x 257 columns]

In [2]:
# split data into training and testing csv's, y is for the target column (int0)
y = data.int0
X = data.drop('int0', axis=1)
X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.2)
pd.concat([X_train, y_train], axis=1).to_csv('data/usps_train.csv',index=False)
pd.concat([X_test, y_test], axis=1).to_csv('data/usps_valid.csv',index=False)

### Installing Brainome via Pip
Simply run the cell below in order to install Brainome and be able to use it in terminal

In [3]:
# ! pip install brainome

## 1. Get Measurements

We always want to measure our data before building our predictor in order to ensure we are building the right model. For more information about how to use Daimensions and why we want to measure our data beforehand, check out the Titanic notebook. Don't forget to use -target int0 because the target column is not on the very right for this dataset.

In [3]:
! brainome -measureonly data/usps_train.csv -target int0


Brainome Table Compiler v1.005-7-prod
Copyright (c) 2019-2021 Brainome, Inc. All Rights Reserved.
Licensed to:                 Alexander Makhratchev  (Evaluation)
Expiration Date:             2021-08-31   26 days left
Maximum File Size:           30 GB
Maximum Instances:           unlimited
Maximum Attributes:          unlimited
Maximum Classes:             unlimited
Connected to:                daimensions.brainome.ai  (local execution)

Command:
    btc -measureonly data/usps_train.csv -target int0

Start Time:                 08/05/2021, 16:51 PDT

Cleaning...done. 
Splitting into training and validation...done. 
Pre-training measurements...done. 


Pre-training Measurements
Data:
    Input:                      data/usps_train.csv
    Target Column:              int0
    Number of instances:       7438
    Number of attributes:       256 out of 256
    Number of classes:           10

Class Balance:                
                               1: 16.94%
                         

## 2. Build the Predictor

Based on our measurements, Daimensions recommends we use a neural network (higher expected generalization) and more effort for this dataset. Don't forget to use -target because the target column isn't on the very right.

In [4]:
! brainome -f NN data/usps_train.csv -o usps_predict.py -target int0 -e 5 --yes


Brainome Table Compiler v1.005-7-prod
Copyright (c) 2019-2021 Brainome, Inc. All Rights Reserved.
Licensed to:                 Alexander Makhratchev  (Evaluation)
Expiration Date:             2021-08-31   26 days left
Maximum File Size:           30 GB
Maximum Instances:           unlimited
Maximum Attributes:          unlimited
Maximum Classes:             unlimited
Connected to:                daimensions.brainome.ai  (local execution)

Command:
    btc -f NN data/usps_train.csv -o usps_predict.py -target int0 -e 5 --yes

Start Time:                 08/05/2021, 16:55 PDT

Cleaning...done. 
Splitting into training and validation...done. 
Pre-training measurements...done. 


Pre-training Measurements
Data:
    Input:                      data/usps_train.csv
    Target Column:              int0
    Number of instances:       7438
    Number of attributes:       256 out of 256
    Number of classes:           10

Class Balance:                
                               1: 16.94%
  

## 3. Validate the Model

Now we can validate our model on our test data, a separate set of data that wasn't used for training.

In [5]:
! python3 usps_predict.py -validate data/usps_valid.csv 

Classifier Type:                    Neural Network
System Type:                        10-way classifier

Accuracy:
    Best-guess accuracy:            15.75%
    Model accuracy:                 94.35% (1755/1860 correct)
    Improvement over best guess:    78.60% (of possible 84.25%)

Model capacity (MEC):               622 bits
Generalization ratio:               9.26 bits/bit

Confusion Matrix:

      Actual |          Predicted           
    --------------------------------------------------
           1 |288   0   1   1   0   0   3   0   0   0
           9 |  0 129   1   4   2   3   0   1   0   5
           3 |  1   3 162   4   0   0   0   0   0   4
           4 |  0   2   6 148   1   0   0   0   0   7
          10 |  1   1   0   0 169   8   0   2   0   0
           8 |  0   0   2   1   3 144   0   1   0   1
           7 |  2   3   1   0   0   0 161   2   2   0
           5 |  0   1   2   0   4   0   3 162   1   0
           2 |  0   2   0   0   0   0   0   0 249   1
           6

Hooray! We have validated the accuracy of our model and found that it has a 94.35% accuracy for the test data. We can also see the confusion matrix, which tells us the percentage of data points from each class (columns) that were predicted to be in a certain class (rows). The diagonals are correctly predicted data points.